clients/src/main/java/org/apache/kafka/common/serialization/ListDeserializer.java

```diff
@@ -0,0 +1,197 @@
+            throw new ConfigException(innerSerdePropertyName, innerSerdeClassOrName, "Deserializer's inner serde class \"" + innerSerdeClassOrName + "\" could not be found.");
+        }
+    }
+
+
+    @SuppressWarnings("unchecked")
+    private List<Inner> getListInstance(int listSize) {
+        try {
+            Constructor<List<Inner>> listConstructor;
+            try {
+                listConstructor = (Constructor<List<Inner>>) listClass.getConstructor(Integer.TYPE);
+                return listConstructor.newInstance(listSize);
+            } catch (NoSuchMethodException e) {
+                listConstructor = (Constructor<List<Inner>>) listClass.getConstructor();
+                return listConstructor.newInstance();
+            }
+        } catch (InstantiationException | IllegalAccessException | NoSuchMethodException |
+                IllegalArgumentException | InvocationTargetException e) {
+            throw new KafkaException("Could not construct a list instance of \"" + listClass.getCanonicalName() + "\"", e);
+        }
+    }
+
+    private SerializationStrategy parseSerializationStrategyFlag(final int serializationStrategyFlag) throws IOException {
+        if (serializationStrategyFlag < 0 || serializationStrategyFlag >= SerializationStrategy.VALUES.length) {
+            throw new SerializationException("Invalid serialization strategy flag value");
+        }
+        return SerializationStrategy.VALUES[serializationStrategyFlag];
+    }
+
+    private List<Integer> deserializeNullIndexList(final DataInputStream dis) throws IOException {
+        int nullIndexListSize = dis.readInt();
+        List<Integer> nullIndexList = new ArrayList<>(nullIndexListSize);
+        while (nullIndexListSize != 0) {
+            nullIndexList.add(dis.readInt());
+            nullIndexListSize--;
+        }
+        return nullIndexList;
+    }
+
+    @Override
+    public List<Inner> deserialize(String topic, byte[] data) {
+        if (data == null) {
+            return null;
+        }
+        try (final DataInputStream dis = new DataInputStream(new ByteArrayInputStream(data))) {
+            SerializationStrategy serStrategy = parseSerializationStrategyFlag(dis.readByte());
+            List<Integer> nullIndexList = null;
+            if (serStrategy == SerializationStrategy.CONSTANT_SIZE) {
+                nullIndexList = deserializeNullIndexList(dis);
+            }
+            final int size = dis.readInt();
+            List<Inner> deserializedList = getListInstance(size);
+            for (int i = 0; i < size; i++) {
+                if (serStrategy == SerializationStrategy.CONSTANT_SIZE
+                        && nullIndexList.contains(i)) {
+                    deserializedList.add(null);
+                    continue;
+                }
+                int entrySize = (primitiveSize == null || serStrategy == SerializationStrategy.VARIABLE_SIZE) ? dis.readInt() : primitiveSize;
```

Comment:
Since we no longer expose the SerializationStrategy or let users explicitly select it, these two equality checks should have both be true or both be false, right? Might read a bit easier if we only check `serStrategy == SerializationStrategy.VARIABLE_SIZE` here, and then just verify that `primitiveSize` is not null when we parse the serialization strategy flag at the top. WDYT?